pkg/controller/prpqr_reconciler/prpqr_reconciler.go

```diff
@@ -0,0 +1,193 @@
+	err := r.reconcile(ctx, req, logger)
+	if err != nil {
+		logger.WithError(err).Error("Reconciliation failed")
+	} else {
+		logger.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, logger *logrus.Entry) error {
+	logger = logger.WithField("namespace", req.Namespace).WithField("prpqr_name", req.Name)
+	logger.Info("Starting reconciliation")
+
+	prpqr := &v1.PullRequestPayloadQualificationRun{}
+	if err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: req.Namespace, Name: req.Name}, prpqr); err != nil {
+		return fmt.Errorf("failed to get the PullRequestPayloadQualificationRun: %s in namespace %s: %w", req.Name, req.Namespace, err)
+	}
+
+	prowjobs := generateProwjobs(prpqr.Spec.PullRequest.Org, prpqr.Spec.PullRequest.Repo, prpqr.Spec.PullRequest.BaseRef, req.Name, req.Namespace)
+	for _, pj := range prowjobs {
+		logger = logger.WithFields(logrus.Fields{"name": pj.Name, "namespace": req.Namespace})
+
+		err := r.client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: pj.Namespace, Name: pj.Name}, &prowv1.ProwJob{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get the PullRequestPayloadQualificationRun: %s in namespace %s: %w", req.Name, req.Namespace, err)
+		}
+
+		if !kerrors.IsNotFound(err) {
+			continue
+		}
+
+		logger.Info("Creating prowjob...")
+		if err := r.client.Create(ctx, &pj); err != nil {
+			return fmt.Errorf("failed to create prowjob: %w", err)
+		}
+
+		// There is some delay until it gets back to our cache, so block until we can retrieve
+		// it successfully.
+		key := ctrlruntimeclient.ObjectKey{Namespace: pj.Namespace, Name: pj.Name}
+		if err := wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+			if err := r.client.Get(ctx, key, &prowv1.ProwJob{}); err != nil {
+				if kerrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, fmt.Errorf("getting prowJob failed: %w", err)
+			}
+			return true, nil
+		}); err != nil {
+			return fmt.Errorf("failed to wait for created ProwJob to appear in cache: %w", err)
+		}
+
+		prpqr.Status.Jobs = append(prpqr.Status.Jobs, v1.PullRequestPayloadJobStatus{
+			ReleaseJobName: pj.Spec.Job,
+			ProwJob:        pj.Name,
+			Status:         pj.Status,
+		})
+
+		logger.Info("Updating PullRequestPayloadQualificationRun...")
+		if err := r.client.Update(ctx, prpqr); err != nil {
+			return fmt.Errorf("failed to update PullRequestPayloadQualificationRun %s: %w", prpqr.Name, err)
+		}
+
+	}
+	return nil
+}
+
+// TODO: Currently we create a single dummy prowjob just for testing. The actual implementation
+// will be introduced in https://issues.redhat.com/browse/DPTP-2577
+func generateProwjobs(org, repo, branch, prpqrName, prpqrNamespace string) []prowv1.ProwJob {
+	labels := map[string]string{
+		v1.PullRequestPayloadQualificationRunLabel: prpqrName,
```

Comment:
label value has some restrictions and we have some lib on it.
Should we ensure the restrictions here?

Reply:
We could, but this is a constant. I can still write a test to ensure if we change the value, we won't break it.